bundles/org.openhab.binding.homekit/src/main/java/org/openhab/binding/homekit/internal/transport/IpTransport.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (c) 2010-2025 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.homekit.internal.transport;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.homekit.internal.session.AsymmetricSessionKeys;
+import org.openhab.binding.homekit.internal.session.SecureSession;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This provides the IP transport layer for HomeKit communication.
+ * It provides methods for sending GET, POST, and PUT requests with appropriate headers and content types.
+ * It supports both plain and secure (encrypted) communication based on whether session keys have been set.
+ * It handles building HTTP requests, sending them over a socket, and parsing HTTP responses.
+ * It throws exceptions for various error conditions, including IO issues, timeouts, and non-200 HTTP responses.
+ *
+ * @author Andrew Fiddian-Green - Initial contribution
+ */
+@NonNullByDefault
+public class IpTransport implements AutoCloseable {
+
+    private static final int CONNECT_TIMEOUT = Duration.ofSeconds(5).toMillisPart();
+
+    private final Logger logger = LoggerFactory.getLogger(IpTransport.class);
+
+    private final String host; // ip address with optional port e.g. "192.168.1.42:9123"
+    private final Socket socket;
+
+    private @Nullable SecureSession secureSession = null;
+
+    /**
+     * Creates a new IpTransport instance with the given socket and session keys.
+     */
+    public IpTransport(String host) throws Exception {
+        logger.debug("Connecting to {}", host);
+        this.host = host;
+        String[] parts = host.split(":");
+        String ipAddress = parts[0];
+        int port = parts.length > 1 ? Integer.parseInt(parts[1]) : 0;
```

Comment:
I guess a port should always be explicitly required (i.e. as parameter for this function and probably also as a separate configuration parameter)? It's not possible to **connect** to port 0, only to ask to **bind** to port 0 to get any port for a listener.

Reply:
Indeed. And you will never know the port unless you use auto- discovery. => Please please do not go any further down the road of manual '.things' based thing creation! It will simply NOT work.

Reply:
As explained [here](https://github.com/openhab/openhab-addons/pull/19340#issuecomment-3353855034) I could easily find the port. If it had been possible for me to use mDNS across my VLAN's I could also have found it through discovery, but then still decided to create a file-based configuration.

However, this is not related to my comment: A hostname is provided with the expected format `<host>[:<port>]`. But if port is omitted, it defaults to 0, which won't work. So in this case you could throw `IllegalArgumentException` or you could instead divide it into two separate parameters: host and port, which would make sure that the port is always provided.

Reply:
I will leave this comment open until we have resolve the question whether or not it should be encouraged for users to attempt to create manual configurations. I prefer to defer the discussion about adding user complications until after we have a binding that works for the simple user case.

Reply:
Ok. Done